### PR TITLE
[fix][txn] fix ack with txn compute ackedCount error

### PR DESCRIPTION
**Pull request by congbobo184**
issue: 
reproduce:
1. redeliver one consumer received message of a batch message, will remove this message from this consumer pending acks
2. if this message is removed and doesn't redeliver to another consumer or this consumer, we ack another message in the same batch
3. to compute the consumer unackedCount will get an incorrect batch size, because the message don't in any consumer pending acks
### Motivation
get correct batch size to compute the consumer unackedCount
### Modification
use the correct batch size from the `messageIdDate`
### Verifying this change
add the test

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

- [x] `doc-not-needed`

### CODE REVIEW

**Review by liangyepianzhou — COMMENTED** [review]
Maybe we can also change `get batchSize` in the method `individualAckNormal`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java [hunk 1]
```diff
@@ -523,25 +523,31 @@ private CompletableFuture<Long> individualAckWithTransaction(CommandAck ack) {
...
             MessageIdData msgId = ack.getMessageIdAt(i);
-            PositionImpl position;
+            PositionImpl position = PositionImpl.get(msgId.getLedgerId(), msgId.getEntryId());
+            // acked count at least one
             long ackedCount = 0;
-            long batchSize = getBatchSize(msgId);
+            long batchSize = 0;
+            if (msgId.hasBatchSize()) {
+                batchSize = msgId.getBatchSize();
+                // ack batch messages set ackeCount = 1, if has ackSets will recompute
```
Comment (gaoran10) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:532]:
```suggestion
                // ack batch messages set ackeCount = batchSize, if has ackSets will recompute
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java [hunk 2]
```diff
@@ -523,25 +523,31 @@ private CompletableFuture<Long> individualAckWithTransaction(CommandAck ack) {
         LongAdder totalAckCount = new LongAdder();
         for (int i = 0; i < ack.getMessageIdsCount(); i++) {
             MessageIdData msgId = ack.getMessageIdAt(i);
-            PositionImpl position;
+            PositionImpl position = PositionImpl.get(msgId.getLedgerId(), msgId.getEntryId());
+            // acked count at least one
             long ackedCount = 0;
-            long batchSize = getBatchSize(msgId);
+            long batchSize = 0;
```
Comment (gaoran10) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:529]: The default value may be 1.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java [hunk 3]
```diff
@@ -523,25 +523,31 @@ private CompletableFuture<Long> individualAckWithTransaction(CommandAck ack) {
...
             if (msgId.getAckSetsCount() > 0) {
                 long[] ackSets = new long[msgId.getAckSetsCount()];
                 for (int j = 0; j < msgId.getAckSetsCount(); j++) {
                     ackSets[j] = msgId.getAckSetAt(j);
                 }
                 position = PositionImpl.get(msgId.getLedgerId(), msgId.getEntryId(), ackSets);
+                // recompute ackedCount
                 ackedCount = getAckedCountForTransactionAck(batchSize, ackSets);
             } else {
                 position = PositionImpl.get(msgId.getLedgerId(), msgId.getEntryId());
```
Comment (gaoran10) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:550]: Maybe we can remove this line.

**Review by gaoran10 — COMMENTED** [review]
Great work! Left some comments.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java (hunk 2 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:529]: if not batch message, the batchSize can not be able to 1

Comment (gaoran10) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:529]: Ok, If not batch message, the batch size should be 0, right?

**Review by gaoran10 — APPROVED** [review]
LGTM

**Review by Technoboy- — COMMENTED** [review]
When `sendMessages`, we put the batch size into `pendingAcks`, so in general, we may get the batch size from it when process ack...

**Comment by congbobo184** [comment]
@Technoboy- transaction ack command will carry the batchSize, if the batchSize is not in pendingAcks(the message redeliver hasn't been sent to any consumer), we will not get the correct batch size. So I think in Transaction ack, we uniformly use command carried batch size is ok

**Comment by Jason918** [comment]
@Technoboy- Ping
@congbobo184 Please resolve the conflicts.

**Comment by Jason918** [comment]
@Technoboy- Ping
@congbobo184 Please resolve the conflicts.

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**